src/Terminal/ui/TerminalActionTimer.tsx

```diff
@@ -0,0 +1,26 @@
+import React, { useEffect } from "react";
+import _ from "lodash";
+import Typography from "@mui/material/Typography";
+
+import { useRerender } from "../../ui/React/hooks";
+import { Terminal } from "../../Terminal";
+import { TerminalProcessEvents } from "../TerminalEvents";
+
+export function TerminalActionTimer(): React.ReactElement {
+  const rerender = useRerender();
+
+  useEffect(() => {
+    const debounced = _.debounce(rerender, 25, { maxWait: 50 });
+    const unsubscribe = TerminalProcessEvents.subscribe(debounced);
+    return () => {
+      debounced.cancel();
+      unsubscribe();
+    };
```

Comment:
```suggestion
import Typography from "@mui/material/Typography";

import { useRerender } from "../../ui/React/hooks";
import { Terminal } from "../../Terminal";
import { TerminalProcessEvents } from "../TerminalEvents";

export function TerminalActionTimer(): React.ReactElement {
  const rerender = useRerender();

  useEffect(() => {
    const unsubscribe = TerminalProcessEvents.subscribe(rerender);
    return unsubscribe;
```
Debounce is not needed here, it's needed for other terminal stuff because they could happen at arbitrarily fast rates (because the trigger is controlled by the player), but here we're only rerendering every time the terminal process function is ran, which should be a fixed time of approximately once every 200ms.

Reply:
Good catch! 👍
I did not even check how often this event is emitted when I created the PR.

It's indeed called approximately every 200ms. I've decided to use the alternative approach since it's cleaner (less code). And we don't need to check if the action is null there because in this case component is already unmounted.

Reply:
> And we don't need to check if the action is null there because in this case component is already unmounted.
we actually do need to do that 😅 - after some testing (with reduced rerender timeouts) I got an error because of that.
wip...

Reply:
@Snarling could you please review the latest version?

Reply:
https://github.com/bitburner-official/bitburner-src/assets/84951833/31acac28-5b83-48eb-932d-655349e343b4

Not sure if this can happen if the main process loop isn't delayed (e.g. by a script keeping the main process busy), but I was able to get this error by running a script that causes delays in the main engine loop and then just backdooring n00dles repeatedly. This will happen any time that the action has ended (so there's no action to make a bar out of) but the actual terminal has not rerendered yet because even though it got the event trigger from the action finishing, it's a debounced event so it won't happen right away.

Reply:
Just saw your update, that should be fine to avoid the issue.

Reply:
yeah, I just made a wrong assumption that setting the action to null will unmount the progress bar before its rerender is called. The main reason here is that action is not a react state - so changing the action value does not trigger any re-renders directly.

Thanks for the review and valuable advice!